File: src/NFTPersaleFollowing/flashbot.py
import asyncio
import json
from web3 import Web3 # type: ignore
from web3.types import HexBytes # type: ignore
from eth_account import Account # type: ignore
from eth_account._utils.legacy_transactions import (
    encode_transaction,
    serializable_unsigned_transaction_from_dict,
)
from flashbots import flashbot # type: ignore
from websocket import create_connection  # type: ignore
import time
from dotenv import load_dotenv # type: ignore
import os

load_dotenv('/Users/richard/Desktop/dev/openspace2/.env')
Alchemy_API_KEY = os.getenv("ALCHEMY_API_KEY")
PrivateKey = os.getenv("PRIVATE_KEY_DEV_2")

sepolia = {
    "chain_id": 11155111,
    "chain_name": "Sepolia",
    "rpc_url": f"https://eth-sepolia.g.alchemy.com/v2/{Alchemy_API_KEY}",
    "ws_url": f"wss://eth-sepolia.g.alchemy.com/v2/{Alchemy_API_KEY}",
    "explorer_url": "https://sepolia.etherscan.io",
    "flashbots_relays": "https://relay-sepolia.flashbots.net",
    "contract_address": "0xa4010fa5a816747f9eba1a60271280beaae28f10",
}
mainnet = {
    "chain_id": 1,
    "chain_name": "Mainnet",
    "rpc_url": f"https://eth-mainnet.g.alchemy.com/v2/{Alchemy_API_KEY}",
    "ws_url": f"wss://eth-mainnet.g.alchemy.com/v2/{Alchemy_API_KEY}",
    "explorer_url": "https://etherscan.io",
    "flashbots_relays": "https://relay.flashbots.net",
    "contract_address": "0xb6907c2747c462834fc25fb4e76b4c0f51a1f536",
}

network = mainnet
network = sepolia
# 设置 Web3 连接到 Sepolia 测试网络
w3 = Web3(Web3.HTTPProvider(network["rpc_url"]))
# WebSocket 提供者
ws_url = network["ws_url"]
# 设置 Flashbots 中继
flashbot_w3 = flashbot(w3, Account.from_key(PrivateKey), network["flashbots_relays"])

print(f"⚠️⚠️⚠️Connected to {network['chain_name']} with chain ID {network['chain_id']}")

# 对于Web3.py v6+，使用新的中间件导入和应用方式
try:
    # 尝试新版API (web3.py v6+)
    from web3.middleware import geth_poa  # 新的导入路径
    w3.middleware_onion.inject(geth_poa.geth_poa_middleware, layer=0)
except ImportError:
    try:
        # 尝试旧版API (web3.py v5)
        from web3.middleware import geth_poa_middleware
        w3.middleware_onion.inject(geth_poa_middleware, layer=0)
    except ImportError:
        print("警告: 无法加载 POA 中间件，可能影响与 Sepolia 的交互")



# 定义 OpenspaceNFT 合约 ABI 和地址
contract_abi = [
    {"inputs": [], "name": "enablePresale", "outputs": [], "stateMutability": "nonpayable", "type": "function"},
    {"inputs": [{"internalType": "uint256", "name": "amount", "type": "uint256"}], "name": "presale", "outputs": [], "stateMutability": "payable", "type": "function"}
]
# 修复: 使用toChecksumAddress将合约地址转换为checksum格式
contract_address = w3.to_checksum_address(network["contract_address"])
contract = w3.eth.contract(address=contract_address, abi=contract_abi)

# 定义所有者地址和买家私钥 - 同样使用checksum格式
owner_address = w3.to_checksum_address("0xE9d2E42129C04f5627f7894aABD422B8a76737aD")
buyer_private_key = PrivateKey  
buyer_address = w3.to_checksum_address(Account.from_key(buyer_private_key).address)

# 计算 enablePresale 函数选择器
enable_presale_selector = Web3.keccak(text="enablePresale()").hex()[:8]

# 创建并签署用户的 presale 交易
amount = 2  # 示例：购买 2 个 NFT
value = amount * 10**14  # 0.0002 Ether
presale_txn = contract.functions.presale(amount).build_transaction({
    'chainId': w3.eth.chain_id,
    'from': buyer_address,
    'maxFeePerGas': 50 * 10**9,  # 示例 gas 费用
    'maxPriorityFeePerGas': 20 * 10**9,
    'nonce': w3.eth.get_transaction_count(buyer_address),
    'value': value,
    'gas': 300000,
})
signed_presale = w3.eth.account.sign_transaction(presale_txn, buyer_private_key)
raw_presale_tx = HexBytes(signed_presale.rawTransaction)
# 将原始交易转换为十六进制字符串
# raw_presale_tx_hex = HexBytes(raw_presale_tx)
# print(f"Tx Type:{raw_presale_tx[0]}, Presale Tx: {raw_presale_tx},{type(raw_presale_tx)}")


# 恢复原始交易
def recover_raw_transaction(tx):
    """Recover raw transaction for replay.
    Inspired by: https://github.com/ethereum/eth-account/blob/1d26f44f6075d6f283aeaeff879f4508c9a228dc/eth_account/_utils/signing.py#L28-L42
    """
    transaction = dict(tx)
    v, r, s = transaction.pop("v"), transaction.pop("r"), transaction.pop("s")
    transaction.pop('blockHash')
    transaction.pop('blockNumber')
    transaction.pop('transactionIndex')
    transaction.pop('from')
    transaction.pop('yParity')
    transaction.pop('gasPrice')
    transaction.pop('hash')
    transaction['data']=transaction.pop('input')
    # print(f"Transaction: {transaction}")
    unsigned_transaction = serializable_unsigned_transaction_from_dict(transaction)
    raw_transaction = HexBytes("0x" + encode_transaction(unsigned_transaction, vrs=(v, r, s)).hex())
    # print(f"Raw Transaction: {raw_transaction}")
    return raw_transaction

# WebSocket 监听新交易
async def listen_pending_transactions():
    ws = create_connection(ws_url)
    subscription = {"jsonrpc": "2.0", "id": 1, "method": "eth_subscribe", 
                    "params": ["alchemy_pendingTransactions",
                    {"fromAddresses":  owner_address, "toAddress": contract_address, 
                     "includeRemoved": False,"hashesOnly": True,},],}
    ws.send(json.dumps(subscription))
    print("Listening for pending transactions...")

    while True:
        message = json.loads(ws.recv())
        if "params" in message and "result" in message["params"]:
            tx_hash = message["params"]["result"]
            print(f"New pending transaction: {tx_hash}")
            # try:
            if True:
                tx = w3.eth.get_transaction(tx_hash)
                
                if (tx and tx['input'].hex()[:8] == enable_presale_selector):
                    # 找到所有者的 enablePresale 交易
                    # raw_enable_tx = w3.provider.make_request("eth_getRawTransactionByHash", [tx_hash])['result']
                    # raw_enable_tx = w3.eth.get_raw_transaction(tx_hash)
                    raw_enable_tx = recover_raw_transaction(tx)
                    if not raw_enable_tx:
                        print("无法获取原始交易，跳过此交易")
                        continue
                    print(f"type: {type(raw_enable_tx)}, Raw Enable Tx: {raw_enable_tx}")

                    # 创建捆绑
                    # bundle_txs = [raw_enable_tx, raw_presale_tx]
                    bundle_txs = [
                    # {"signed_transaction": raw_enable_tx},
                    {"signed_transaction": raw_presale_tx},
                    ]
                    # bundle_txs = [raw_presale_tx]

                    current_block = w3.eth.block_number
                    # target_block = current_block + 1
                    # target_block_hex = hex(target_block)
                    print(f"Bundle Tx: {bundle_txs}")
                    # print(f"Type: {type(target_block)}, Target Block: {target_block}")

                    # 发送捆绑
                    send_result = flashbot_w3.flashbots.send_bundle(bundle_txs, target_block_number=current_block + 10)
                    bundle_hash = w3.to_hex(send_result.bundle_hash())
                    print(f"Bundle hash: {bundle_hash}")

                    # 获取捆绑状态
                    # stats = await flashbots.get_bundle_stats(bundle_hash)
                    stats = w3.flashbots.get_bundle_stats(bundle_hash, current_block + 10)

                    # 打印结果链接
                    print(f"Enable Presale Tx: {network['explorer_url']}/tx/{tx_hash}")
                    print(f"Presale Tx: {network['explorer_url']}/tx/{signed_presale['hash'].hex()}")
                    print(f"Bundle Stats: {stats}")

                    ws.close()
                    break
            # except Exception as e:
                # print(f"Error processing tx {tx_hash}: {e}")
        await asyncio.sleep(0.1)  # 避免阻塞

# 运行异步监听
if __name__ == "__main__":
    asyncio.run(listen_pending_transactions())

File: src/NFTPersaleFollowing/flashbot.js
import { ethers } from "ethers";
import { FlashbotsBundleProvider } from "@flashbots/ethers-provider-bundle";
import * as dotenv from "dotenv";
import * as path from "path";
import { fileURLToPath } from "url";
import WebSocket from "ws"; // 直接使用ws模块

// 加载.env文件(从上级目录)
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.resolve(__dirname, '../../.env') });

// Alchemy API URL更新
const mainnet = {
  rpc: `https://eth-mainnet.g.alchemy.com/v2/${process.env.ALCHEMY_API_KEY}`,
  // 正确的WebSocket格式
  ws: `wss://eth-mainnet.g.alchemy.com/v2/${process.env.ALCHEMY_API_KEY}`,
  flashbots: "https://relay.flashbots.net"
};

const sepolia = {
  rpc: `https://eth-sepolia.g.alchemy.com/v2/${process.env.ALCHEMY_API_KEY}`,
  // 正确的WebSocket格式 - .ws.alchemyapi.io 改为 .g.alchemy.com
  ws: `wss://eth-sepolia.g.alchemy.com/v2/${process.env.ALCHEMY_API_KEY}`,
  flashbots: "https://relay-sepolia.flashbots.net"
};

// 获取环境变量
const NETWORK = "sepolia";
const CONTRACT_ADDRESS = "0xa4010fa5a816747f9eba1a60271280beaae28f10";
const OWNER_ADDRESS = "0xE9d2E42129C04f5627f7894aABD422B8a76737aD";
const PRIVATE_KEY = process.env.PRIVATE_KEY;
const ETHEREUM_RPC_URL = NETWORK === "mainnet" ? mainnet.rpc : sepolia.rpc;
const FLASHBOTS_RELAY_URL = NETWORK === "mainnet" ? mainnet.flashbots : sepolia.flashbots;
const WS_URL = NETWORK === "mainnet" ? mainnet.ws : sepolia.ws;

// 验证环境变量
if (!process.env.ALCHEMY_API_KEY) {
  console.error("错误: 缺少ALCHEMY_API_KEY。请在.env文件中设置");
  process.exit(1);
}

if (!PRIVATE_KEY) {
  console.error("错误: 缺少私钥。请在.env文件中设置PRIVATE_KEY");
  process.exit(1);
}

// 手动WebSocket管理
async function setupWebSocketListener(url, callback) {
  return new Promise((resolve, reject) => {
    try {
      console.log(`连接到WebSocket: ${url}`);
      const ws = new WebSocket(url);
      
      ws.on('open', () => {
        console.log('WebSocket连接已建立');
        
        // 发送订阅请求
        const subscription = {
          "jsonrpc": "2.0", 
          "id": 1, 
          "method": "eth_subscribe", 
          "params": [
            "alchemy_pendingTransactions",
            {
              // "fromAddress": OWNER_ADDRESS,
              // "toAddress": CONTRACT_ADDRESS,
              "hashesOnly": true
            }
          ]
        };
        
        ws.send(JSON.stringify(subscription));
        console.log('已发送订阅请求');
        
        resolve(ws);
      });
      
      ws.on('message', (data) => {
        try {
          const message = JSON.parse(data.toString());
          callback(message);
        } catch (e) {
          console.error('解析消息失败:', e);
        }
      });
      
      ws.on('error', (error) => {
        console.error('WebSocket错误:', error);
      });
      
      ws.on('close', () => {
        console.log('WebSocket连接已关闭');
      });
      
    } catch (error) {
      console.error('设置WebSocket失败:', error);
      reject(error);
    }
  });
}

async function main() {
  // 连接到以太坊RPC
  console.log("连接到以太坊RPC:", ETHEREUM_RPC_URL);
  const httpProvider = new ethers.providers.JsonRpcProvider(ETHEREUM_RPC_URL);

  // 创建钱包
  const wallet = new ethers.Wallet(PRIVATE_KEY, httpProvider);
  console.log("钱包地址:", wallet.address);
  
  // 获取网络信息
  const network = await httpProvider.getNetwork();
  console.log("连接到网络:", network.name, "链ID:", network.chainId);
    
  // enablePresale 函数选择器
  const enablePresaleSelector = ethers.utils.id('enablePresale()').slice(0, 10);
  console.log("enablePresale 函数选择器:", enablePresaleSelector);

  try {
    // 创建Flashbots provider
    console.log("连接到Flashbots中继:", FLASHBOTS_RELAY_URL);
    const flashbotsProvider = await FlashbotsBundleProvider.create(
      httpProvider, 
      wallet, 
      FLASHBOTS_RELAY_URL
    );
    console.log("Flashbots httpProvider创建成功");

    // 设置WebSocket监听
    const wsListener = await setupWebSocketListener(WS_URL, (message) => {
      // 处理WebSocket消息
      if (message.params && message.params.result) {
        const txHash = message.params.result;
        console.log("检测到新交易:", txHash);
        
        // 这里可以添加交易处理逻辑
        processTransaction(txHash, httpProvider, flashbotsProvider, wallet);
      }
    }).catch(error => {
      console.error("WebSocket设置失败，使用轮询模式作为备份");
      // 可以实现一个备份的轮询机制
    });
    
    // 示范Flashbots bundle (可选)
    await demoFlashbotsBundle(httpProvider, flashbotsProvider, wallet, network);
    
    console.log("\n程序继续运行以监听交易...");
    // 保持程序运行
    await new Promise((resolve) => setTimeout(resolve, 24 * 60 * 60 * 1000)); // 24小时
    
  } catch (error) {
    console.error("Flashbots操作失败:", error);
  }
}

// 处理检测到的交易
async function processTransaction(txHash, provider, flashbotsProvider, wallet) {
  try {
    const tx = await provider.getTransaction(txHash);
    console.log("处理交易:", tx);
    if (!tx) return;
    
    // 检查是否是enablePresale交易
    if (tx.data && tx.data.startsWith("0xa8eac4")) {
      console.log("找到enablePresale交易!");
      
      // 准备presale交易
      const nonce = await provider.getTransactionCount(wallet.address);
      const presaleTx = {
        to: tx.to,
        data: "0x03825e4e0000000000000000000000000000000000000000000000000000000000000001", // presale(1)
        value: ethers.utils.parseEther("0.01"),
        gasLimit: 200000,
        maxFeePerGas: ethers.utils.parseUnits("10", "gwei"),
        maxPriorityFeePerGas: ethers.utils.parseUnits("5", "gwei"),
        nonce: nonce,
        type: 2,
        chainId: tx.chainId
      };
      
      // 发送Flashbots bundle
      await sendFlashbotsBundle(provider, flashbotsProvider, wallet, tx, presaleTx);
    }
  } catch (error) {
    console.error("处理交易失败:", error);
  }
}

// 发送Flashbots bundle
async function sendFlashbotsBundle(provider, flashbotsProvider, wallet, enableTx, presaleTx) {
  try {
    const targetBlock = await provider.getBlockNumber() + 1;
    console.log(`准备发送bundle到区块: ${targetBlock}`);
    
    // 签名bundle
    const signedBundle = await flashbotsProvider.signBundle([
      { 
        signer: wallet,
        transaction: presaleTx
      }
    ]);
    
    const bundleReceipt = await flashbotsProvider.sendRawBundle(
      signedBundle,
      targetBlock
    );
    
    console.log("Bundle已发送, Hash:", bundleReceipt.bundleHash);
    
    // 等待结果
    const waitResponse = await Promise.race([
      bundleReceipt.wait(),
      new Promise((resolve) => setTimeout(() => resolve("超时"), 30000))
    ]);
    
    if (waitResponse === "超时") {
      console.log("Bundle等待超时");
    } else {
      console.log("Bundle结果:", waitResponse);
    }
    
  } catch (error) {
    console.error("发送bundle失败:", error);
  }
}

// 示范Flashbots bundle
async function demoFlashbotsBundle(provider, flashbotsProvider, wallet, network) {
  const nonce = await provider.getTransactionCount(wallet.address);
  const targetBlock = await provider.getBlockNumber() + 1;
  
  console.log("当前区块:", await provider.getBlockNumber());
  console.log("目标区块:", targetBlock);
  
  // 创建一个示例交易
  const exampleTx = {
    to: wallet.address,  // 发送给自己避免浪费ETH
    value: ethers.utils.parseUnits("0.0000001", "ether"),
    gasLimit: 21000,
    maxFeePerGas: ethers.utils.parseUnits("10", "gwei"),
    maxPriorityFeePerGas: ethers.utils.parseUnits("5", "gwei"),
    nonce: nonce,
    type: 2,
    chainId: network.chainId
  };
  
  console.log("示例交易已准备");
  
  try {
    // 签名bundle
    const signedBundle = await flashbotsProvider.signBundle([
      { 
        signer: wallet,
        transaction: exampleTx
      }
    ]);
    
    // 发送bundle
    const bundleReceipt = await flashbotsProvider.sendRawBundle(
      signedBundle,
      targetBlock
    );
    
    console.log("示例Bundle发送结果:", bundleReceipt.bundleHash);
  } catch (error) {
    console.error("示例Bundle发送失败:", error);
  }
}

// 执行主函数
main().catch((error) => {
  console.error("执行失败:", error);
  process.exit(1);
});